tests/components/subaru/test_diagnostics.py

```diff
@@ -0,0 +1,73 @@
+"""Test Subaru diagnostics."""
+import json
+
+import pytest
+
+from homeassistant.components.subaru.const import DOMAIN, ENTRY_COORDINATOR
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import device_registry as dr
+
+from .api_responses import TEST_VIN_2_EV
+
+from tests.common import load_fixture
+from tests.components.diagnostics import (
+    get_diagnostics_for_config_entry,
+    get_diagnostics_for_device,
+)
+
+
+async def test_config_entry_diagnostics(hass: HomeAssistant, hass_client, ev_entry):
+    """Test config entry diagnostics."""
+    assert hass.data[DOMAIN]
+
+    config_entry = hass.config_entries.async_entries(DOMAIN)[0]
+
+    diagnostics_fixture = json.loads(
+        load_fixture("subaru/diagnostics_config_entry.json")
+    )
+
+    assert (
+        await get_diagnostics_for_config_entry(hass, hass_client, config_entry)
+        == diagnostics_fixture
+    )
+
+
+async def test_device_diagnostics(hass: HomeAssistant, hass_client, ev_entry):
+    """Test device diagnostics."""
+    assert hass.data[DOMAIN]
+
+    config_entry = hass.config_entries.async_entries(DOMAIN)[0]
+
+    device_registry = dr.async_get(hass)
+    reg_device = device_registry.async_get_device(
+        identifiers={(DOMAIN, TEST_VIN_2_EV)},
+    )
+    assert reg_device is not None
+
+    diagnostics_fixture = json.loads(load_fixture("subaru/diagnostics_device.json"))
+
+    assert (
+        await get_diagnostics_for_device(hass, hass_client, config_entry, reg_device)
+        == diagnostics_fixture
+    )
+
+
+async def test_device_diagnostics_vehicle_not_found(
+    hass: HomeAssistant, hass_client, ev_entry
+):
+    """Test device diagnostics when the vehicle cannot be found."""
+    assert hass.data[DOMAIN]
+
+    config_entry = hass.config_entries.async_entries(DOMAIN)[0]
+
+    device_registry = dr.async_get(hass)
+    reg_device = device_registry.async_get_device(
+        identifiers={(DOMAIN, TEST_VIN_2_EV)},
+    )
+    assert reg_device is not None
+
+    # Remove vehicle info from hass.data so that vehicle will not be found
+    hass.data[DOMAIN][config_entry.entry_id][ENTRY_COORDINATOR].data.pop(TEST_VIN_2_EV)
```

Comment:
https://developers.home-assistant.io/docs/development_testing/?_highlight=writing&_highlight=tests#writing-tests-for-integrations

Make sure to not interact with any integration details in tests of integrations: Please don't modify `hass.data` in tests.  Instead patch your library